### ISSUE TITLE
Internal patchIso8601Timestamp() isn't safe for 5 digit years

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12547]
For simplicity and speed, the internal `patchIso8601Timestamp()` utility was implemented with a hard-coded 4-digit year in mind. Though, ISO 8601 supports 5-digit years as well. We need to take this into account. Current implementation:

```java
private static final String patchIso8601Timestamp(String string, boolean t) {
    if (string.length() > 11)
        if (t && string.charAt(10) == ' ')
            return string.substring(0, 10) + "T" + string.substring(11);
        else if (!t && string.charAt(10) == 'T')
            return string.substring(0, 10) + " " + string.substring(11);

    return string;
}
```

**Comment by GeorgiMinkov** [comment]
Dear Lucas,

Do you know if this ticket was taken? If not I will like to give it a try as a first issue for the jOOQ community.

Additional question - the _"Convert"_ class was migrated to the org.jooq.impl from version 3.15.0, do we need to update the org.jooq.tools implementation of _"Convert"_.

What is the timeline for this update?

**Comment by lukaseder** [comment]
Thanks for your offer @GeorgiMinkov. This is a very low priority issue. You can only fix the very obvious parts in this method, but you don't have access to the integration tests where the bulk of this work will take place (make sure this works correctly in all dialects).

While I appreciate the offer, I prefer you don't fix this right now, given the low priority and the effect this has on the roadmap's re-prioritisation, which I think isn't worth it right now (until a significant amount of users have this problem).

**Comment by GeorgiMinkov** [comment]
Thank you for your answer @lukaseder. Several questions for you:

- Is there a board or a list in which I can track what task needs to be implemented?
- If there is no such list, what is the level of priority that you are aiming?
- Is there a group chat in which you discuss different problems and approaches? (like Discord, Google Chat)
- About the current one:
-- As this converter by design is "less type-safety", we can update it to be universal between the two ISO upgrades
-- Example:

`

        private static final String patchIso8601Timestamp(String string, boolean t) { 
            return t ? string.contains("T")
                        ? string
                        : string.replaceFirst(" ", "T")
                    : string.replaceFirst("T", " ");
`
Of course this logic can be bound with the specific indexes

If you are satisfied with the above approach I can create a corresponding branch and you can use it (merge it) whenever you need to.

In the mean time I will search for better tasks

**Comment by lukaseder** [comment]
I think the expectations here are off.

This is mainly a commercial product, not really looking for help. I'd rather not engage in the very time consuming tasks of project management for marginally beneficial improvements like this one (super low prio), nor for big tasks, which you can't really integration tests (tests aren't open source).

You'll find the official take on the above here, too: https://github.com/jOOQ/jOOQ/blob/main/CONTRIBUTING.md

To answer your specific questions:

> Is there a board or a list in which I can track what task needs to be implemented?

This issue tracker is it, but the priorities aren't publicly tracked. jOOQ 3.16 will be all about:

- The `SQLDialect.YUGABYTE` dialect
- Spatial extensions
- The new query object model

All of these tasks are too vast for external help. You could only ever implement 10% of a task, without access to the closed source tooling (e.g. tests, API generators, etc.)

> If there is no such list, what is the level of priority that you are aiming?

There's a high level roadmap here: https://github.com/jOOQ/jOOQ/projects, but it's not authoritative, nor very useful for external contributors, if there's any.

> Is there a group chat in which you discuss different problems and approaches? (like Discord, Google Chat)

Synchronous chats mean you get premium access to me, and the expectation is that this should be free and instant. I don't even offer this premium level SLA to jOOQ Enterprise Edition customers, who get a guaranteed email reply within 1 business day.

Why do you need it? Why isn't this issue tracker good enough?

> About the current one: As this converter by design is "less type-safety", we can update it to be universal between the two ISO upgrades

Yes, we could. The motivation to implement it as it is now is that this is faster (in benchmarks!), because we don't have to run contains which loops over the entire string every time just to detect a `T`.

Now, we could argue if this is premature optimisation or not. It wasn't for some users back some years ago, who discovered quite a few bottlenecks in this area. And quickly, you can see how time consuming such a trivial task starts out to be. It's not worth it (yet) for a task that probably no one has run into. Yes, it's incorrect, but not *every* bug needs to be addressed immediately. There are a lot of RDBMS out there that can't handle 5-digit years. So, what will we do about those? Are we even ready to spend all this time thinking about such a silly problem that won't be relevant for another 8000 years? Or should I rather invest time in the above 3 high demand features?

> If you are satisfied with the above approach I can create a corresponding branch and you can use it (merge it) whenever you need to.

No, please don't :) I don't want to prioritise this task just because you're offering help. The PR will be rejected.

> In the mean time I will search for better tasks

But why? What's the motivation for this?

**Comment by GeorgiMinkov** [comment]
Thank you for your detailed answers, Lucas.

> Why do you need it? Why isn't this issue tracker good enough?
As various open-source projects have group chats in which collaborators can discuss solutions and work, I found out that this is a good way to maintain a high code quality level. But the tracker is still good.

> But why? What's the motivation for this?
(I will try to keep it short) Interesting question, as for several years I'm using jOOQ and really appreciate your work over the years. I thought that returning the favor is only normal. Of course, like many people inside the community, I'm a fan of open source and continuous improvements.

**Comment by lukaseder** [comment]
> As various open-source projects have group chats in which collaborators can discuss solutions and work, I found out that this is a good way to maintain a high code quality level.

It's a matter of taste. In my personal opinion, synchronous communication mainly produces busywork, and not necessarily tangible results. I don't really see a connection to code quality.

> I thought that returning the favor is only normal.

The best way to do that is to go here :)
https://www.jooq.org/download/price-plans

**Comment by lukaseder** [comment]
So, despite my efforts to explain why I prefer not to get a PR for this seemingly trivial, yet non-trivial issue here, there was a PR 😅 https://github.com/jOOQ/jOOQ/pull/12705, and it was flawed, just creating busywork for something that really isn't top prio.

In addition to the above, some new facts that no ad-hoc contributor could have foreseen are now uncovered: There are new plans of fixing this much more thoroughly as related discussions have shown in:

- https://github.com/jOOQ/jOOQ/issues/12699
- https://github.com/jOOQ/jOOQ/issues/12703

This proves once more that ad-hoc contributors don't have the big picture.

I'm sorry that Github suggests that PRs are a thing. They're not in this particular repository. I have tried to get Github to deactivate this feature (without deactivating the issue tracker) for years now, no success...

**Comment by lukaseder** [comment]
Fixed for jOOQ 3.19.0

Like https://github.com/jOOQ/jOOQ/issues/13786, I won't backport this due to the regression risk in string parsing, which is always hard, and prone to errors due to edge cases. Given that 5 digit years (or 3 digit years) aren't too popular in most data sets, I don't think this is would be an important backport.

**Comment by lukaseder** [comment]
Will look into the other linked issues this afternoon:


- https://github.com/jOOQ/jOOQ/issues/12699
- https://github.com/jOOQ/jOOQ/issues/12703
